### DISCUSSION TITLE
`settings.<package>.custom` is ignored when `package.<package>.source` is a git repository

### DISCUSSION THREAD

**Discussion by Thrithralas** [srid/haskell-flake#237] [Q&A]
## Problem

When specifying source as per [the documentation](https://community.flake.parts/haskell-flake/dependency#path), `settings.<package>.custom` is *seemingly* not applied.

### Detailed Description

I'm not sure if this is actually a bug, I'm probably doing something wrong. Specifically, I'm trying to make a project with `llvm-hs` as its dependency, but I'm trying to use the `llvm-15` branch from the git repository, as the hackage package is extremely outdated. My solution to adding in the llvm dependency looked like the following in when using the hackage version (some settings omitted for clarity, but I can post the whole file if need be):
```nix
{
  inputs = {
    nixpkgs.url = "github:nixos/nixpkgs/nixpkgs-unstable";
    flake-parts.url = "github:hercules-ci/flake-parts";
    haskell-flake.url = "github:srid/haskell-flake";
    ...
  };
  outputs = inputs@{ self, nixpkgs, flake-parts, ... }:
    flake-parts.lib.mkFlake { inherit inputs; } {
      systems = nixpkgs.lib.systems.flakeExposed;
      .
      .
      .
      perSystem = { self', pkgs, config, ... }:
        {
          haskellProjects.default = {
            settings.llvm-hs = {
              broken = false;
              jailbreak = true;
              custom = pkgs.haskell.lib.compose.overrideCabal (drv: {
                libraryToolDepends = (drv.libraryToolDepends or []) ++ [ pkgs.llvmPackages_9.llvm.dev ];
              });
            };
           .
           .
           .
          };
       };
    };
}

```
This works fine and it finds `llvm-config` (it runs into another unrelated error after it, but that's irrelevant)

The same however does not work, the moment I switch I specify source to be the git repository:
```nix
{
  inputs = {
    nixpkgs.url = "github:nixos/nixpkgs/nixpkgs-unstable";
    flake-parts.url = "github:hercules-ci/flake-parts";
    haskell-flake.url = "github:srid/haskell-flake";
    llvm-hs = {
      url = "github:llvm-hs/llvm-hs/llvm-15";
      flake = false;
    };
    ...
  };
  outputs = inputs@{ self, nixpkgs, flake-parts, ... }:
    flake-parts.lib.mkFlake { inherit inputs; } {
      systems = nixpkgs.lib.systems.flakeExposed;
      .
      .
      .
      perSystem = { self', pkgs, config, ... }:
        {
          haskellProjects.default = {
            packages.llvm-hs.source = inputs.llvm-hs + "/llvm-hs"; # the project is not in the root directory
            settings.llvm-hs = {
              broken = false;
              jailbreak = true;
              custom = pkgs.haskell.lib.compose.overrideCabal (drv: {
                libraryToolDepends = (drv.libraryToolDepends or []) ++ [ pkgs.llvmPackages_15.llvm.dev ];
              });
            };
           .
           .
           .
          };
       };
    };
}
```
The build now fails, due it not being able to find `llvm-config` on the path. The package does supply the `llvm-config` program, as entering a nix shell with the program installed and just building the `llvm-hs` package with cabal is working fine. Same issue with the `llvm-12` branch. The `llvm-9` branch has a different error, where `llvm-config` is explicitly written into the cabal file as an external dependency and it fails due to an unsupplied module parameter.

I'll add a minimal reproducible example in a few hours, once I get home.

The error, just in case it's necessary:
```
error: builder for '/nix/store/fvw0fsqpv5ikg6d6frjgphhwsgs6pxaw-llvm-hs-source-15.0.0.drv' failed with exit code 1;
       last 10 log lines:
       > [2 of 2] Linking Setup
       > Running phase: updateAutotoolsGnuConfigScriptsPhase
       > Running phase: configurePhase
       > configureFlags: --verbose --prefix=/nix/store/vnc6pvs123g7ihcgxbhn2rw1mhbksbz9-llvm-hs-source-15.0.0 --libdir=$prefix/lib/$compiler --libsubdir=$abi/$libname --docdir=/share/doc/llvm-hs-15.0.0 --with-gcc=gcc --package-db=/build/tmp.twmXnmS6dC/package.conf.d --ghc-options=-j16 +RTS -A64M -RTS --disable-split-objs --enable-library-profiling --profiling-detail=exported-functions --disable-profiling --enable-shared --disable-coverage --enable-static --disable-executable-dynamic --enable-tests --disable-benchmarks --enable-library-vanilla --disable-library-for-ghci --ghc-option=-split-sections --ghc-options=-haddock --extra-lib-dirs=/nix/store/wyw0d6r22mzl4g7wk1rkif9fsks00qbv-ncurses-6.4/lib --extra-lib-dirs=/nix/store/zabxhfpsgkb9c4sb7fy50pn1l1kczzv2-libffi-3.4.4/lib --extra-lib-dirs=/nix/store/wanbfij8izz6kjxrqj21gh2fdcb93wg7-gmp-with-cxx-6.3.0/lib
       > Using Parsec parser
       > CallStack (from HasCallStack):
       >   withMetadata, called at libraries/Cabal/Cabal/src/Distribution/Simple/Utils.hs:370:14 in Cabal-3.8.1.0:Distribution.Simple.Utils
       > Error: Setup: The program 'llvm-config' version >=15.0 && <15.1 is required
       > but it could not be found.
       >
       For full logs, run 'nix log /nix/store/fvw0fsqpv5ikg6d6frjgphhwsgs6pxaw-llvm-hs-source-15.0.0.drv'.
```

**Comment by srid** [comment]
The `settings` overlay is applied on top of the `packages` overlay:

https://github.com/srid/haskell-flake/blob/1e297173b23c5113dd90a2d299d6e0d864af35f1/nix/modules/project/outputs.nix#L86-L89

So I don't see how one could impact the other.

Are you able to reproduce this with some other simple package (not llvm-hs). Because, maybe the git repo is at fault here for not respecting `libraryToolDepends`?

But yes, if you have a minimal repro, please file a bug report. I've made this one a Discussion since we don't know for sure yet whether this is a bug in haskell-flake.

**Reply by Thrithralas** [reply]
I've made a minimum reproducible example, see here: https://github.com/Thrithralas/haskell-flake-237-reproducible
- The `llvm` directory contains a reproducible setup for `llvm-hs` specifically. Uncommenting the source declaration and the and changing the llvm version to 15 produces the issue as shown in its README. Without the source, it builds fine, but crashes later on (though in the error, the succesfull use of llvm-config can be seen).
- The `sample-project` is a project which uses the `hello` program in the same manner as `llvm-hs` uses `llvm-configure` in its Setup phase.
- `using-sample-project` simply uses the project above, failing to find `hello` during the Setup phase.

I'll try to add a reproducibe build for an existing package once I get home. I'm not that familiar with the haskell/nix infrastructure, but `libraryToolDepends` seems to be for packages which are required during the Setup phase.

**Reply by johnhampton** [reply]
I ran into this problem as well. I was planning on filing a bug report today. It appears that none of the settings i.e. `check` and `jailbreak` are applied when the source is a specified. It broke fairly recently, commit c301f31092a74a6acde9d98d6431ea0a9952ca0f works fine but master doesn't work. 

https://github.com/johnhampton/haskell-flake-bug-repo demonstrates the problem.

**Reply by srid** [reply]
Thanks, let me try to reproduce this. John's repo is taking a while to build. 

https://github.com/srid/haskell-flake/commit/dd0d3b9d51f226be15367bf6c90006bd0f859f06 is the only commit I see to be significant recent change.

**Reply by srid** [reply]
Confirmed: https://github.com/srid/haskell-flake/issues/238
